### PR TITLE
Reference super features in assembly module

### PR DESCRIPTION
**Pull request by neykov**
Depends on https://github.com/jclouds/jclouds-karaf/pull/98

### CODE REVIEW

**Comment by andreaturli** [comment]
@neykov I assume this will work once we merge jclouds-karaf#98 right?

File: assembly/pom.xml [hunk 1]
```diff
@@ -174,63 +174,19 @@
...
+                <descriptor>
+                  mvn:org.apache.jclouds.karaf/jclouds-karaf-labs/${jclouds.karaf.version}/xml/features
+                </descriptor>
               </descriptors>
               <features>
                 <feature>config</feature>
-                <feature>jclouds</feature>
-                <feature>jclouds-commands</feature>
-                <feature>jclouds-compute</feature>
-                <feature>jclouds-guice</feature>
```
Comment (demobox) [inline assembly/pom.xml]: Do we need to keep these? They are not directly included by the macro features, from what I can see?

Even if they don't _need_ to be included, would it be clearer to keep them?

**Comment by demobox** [comment]
@neykov Thanks for taking care of this! One thing: since I suspect the tests for the CLI are not entirely conclusive, please package a distribution from this branch and test it locally.

**Comment by neykov** [comment]
@andreaturli This doesn't need `jclouds-karaf#98`. It's just for backwards compatibility.
@demobox I've tried the distribution with `bin/jclouds` - it's working as expected.

File: assembly/pom.xml (hunk 1 above)
Comment (neykov) [inline assembly/pom.xml]: `jclouds-guice`, `jclouds` is included by `jclouds-compute` and it is included by all the compute features. `jclouds-commands` is kept further below. It's like maven transitive features, don't need to mention them if we don't directly depend on them.

**Comment by andreaturli** [comment]
lgtm

**Comment by neykov** [comment]
Correction on my comment above @andreaturli. This does require https://github.com/jclouds/jclouds-karaf/pull/98 to build successfully. And https://github.com/jclouds/jclouds-karaf/pull/99 is not needed, it's just good to have.

**Comment by demobox** [comment]
> I've tried the distribution with bin/jclouds - it's working as expected.

Ah, great - thanks for the additional testing!

File: assembly/pom.xml (hunk 1 above)
Comment (demobox) [inline assembly/pom.xml]: > It's like maven transitive features, don't need to mention them if we don't directly depend on them.

Thanks for clarifying. I suspected as much, but I think we may be specifying similar features elsewhere - not sure why, perhaps for clarity?

But it's a minor comment - just wondering what the most consistent approach for this refactoring is, that's all.

**Comment by demobox** [comment]
Closing and reopening to trigger a rebuild

**Comment by demobox** [comment]
> 1 failing check

```
[ERROR] Failed to execute goal org.apache.karaf.tooling:features-maven-plugin:2.3.11:add-features-to-repo (add-features-to-repo) on project jclouds-cli-assembly: Error populating repository: Unable to find the feature 'jclouds-labs-all-compute' -> [Help 1]
```

I think this may be because the [jclouds-karaf build](https://jclouds.ci.cloudbees.com/job/jclouds-karaf/2124/) for the merge of jclouds/jclouds-karaf#98 is still running. Will try again once that's completed.

**Comment by demobox** [comment]
retest this please

**Comment by demobox** [comment]
> 1 failing check

Hrm...now failing with the following?

```
[ERROR] Failed to execute goal org.apache.karaf.tooling:features-maven-plugin:2.3.11:add-features-to-repo (add-features-to-repo) on project jclouds-cli-assembly: Can't resolve bundle org.apache.jclouds.karaf:jclouds-karaf:cfg:shell:2.1.0-SNAPSHOT: Could not find artifact org.apache.jclouds.karaf:jclouds-karaf:cfg:shell:2.1.0-SNAPSHOT in cloudbees-private-snapshot-
```

/cc @neykov @andreaturli @nacx Any thoughts on where that might be coming from?

**Comment by demobox** [comment]
> Can't resolve bundle org.apache.jclouds.karaf:jclouds-karaf:cfg:shell:2.1.0-SNAPSHOT: Could not find artifact org.apache.jclouds.karaf:jclouds-karaf:cfg:shell:2.1.0-SNAPSHOT

Hm, so we are indeed publishing less to snapshots via the Jenkins build than we are during (I suspect) the release process. Compare [the 2.0.1 release](https://repository.apache.org/content/groups/public/org/apache/jclouds/karaf/jclouds-karaf/2.0.1/):

```
Parent Directory
jclouds-karaf-2.0.1-credentials.cfg	Wed Feb 22 11:33:38 UTC 2017	823	
...
jclouds-karaf-2.0.1-features.xml	Wed Feb 22 11:33:35 UTC 2017	28458	
...
jclouds-karaf-2.0.1-shell.cfg	Wed Feb 22 11:33:39 UTC 2017	2270	
...	
jclouds-karaf-2.0.1.pom	Wed Feb 22 11:33:31 UTC 2017	7502	
...
```

with [2.1.0-SNAPSHOT](https://repository.apache.org/content/groups/snapshots/org/apache/jclouds/karaf/jclouds-karaf/2.1.0-SNAPSHOT/):

```
jclouds-karaf-2.1.0-20170320.011813-88-features.xml	Mon Mar 20 01:18:14 UTC 2017	43203	
...	
jclouds-karaf-2.1.0-20170320.011813-88.pom	Mon Mar 20 01:18:13 UTC 2017	1670	
...
maven-metadata.xml	Mon Mar 20 01:18:14 UTC 2017	832	
maven-metadata.xml.md5	Mon Mar 20 01:46:47 UTC 2017	33	
maven-metadata.xml.sha1	Mon Mar 20 01:46:47 UTC 2017	41	
```

**Comment by nacx** [comment]
Looks like we need to bring this back to the features pom.xml in jclouds-karaf?
https://github.com/jclouds/jclouds-karaf/blob/2.0.x/feature/pom.xml#L34-L91

**Comment by neykov** [comment]
Nice find @demobox, @nacx - added it back in https://github.com/jclouds/jclouds-karaf/pull/100.

**Comment by neykov** [comment]
Tested successfully on top of both https://github.com/jclouds/jclouds-karaf/pull/100 and https://github.com/jclouds/jclouds-karaf/pull/99.

**Comment by nacx** [comment]
jclouds-karaf snapshots should be now in place.
rebuild please

**Comment by nacx** [comment]
rebuild please

**Comment by nacx** [comment]
Looks like something is not yet OK with all these changes. I've build this branch, and after entering the CLI in interactive mode there are no features available:

```
   _       _                 _      
  (_)     | |               | |     
   _  ____| | ___  _   _  _ | | ___ 
  | |/ ___) |/ _ \| | | |/ || |/___)
  | ( (___| | |_| | |_| ( (_| |___ |
 _| |\____)_|\___/ \____|\____(___/ 
(__/                                 

  Apache jclouds cli (2.1.0-SNAPSHOT)
  https://jclouds.apache.org

Hit '<tab>' for a list of available commands
and '[cmd] --help' for help on a specific command.
Hit '<ctrl-d>' to shutdown jclouds cli.

jclouds> features:list
No features available.
```

**Comment by neykov** [comment]
Haven't investigated why `bin/jclouds` works and `bin/jclouds-cli` doesn't. The error with `bin/jclouds-cli` is:
```
2017-03-21 14:17:23,705 | WARN  | FelixStartLevel  | FeaturesServiceImpl              | res.internal.FeaturesServiceImpl 1130 | 21 - org.apache.karaf.features.core - 2.3.11 | Unable to add features repository mvn:org.apache.jclouds.karaf/jclouds-karaf/2.1.0-SNAPSHOT/xml/features at startup
java.lang.IllegalArgumentException: Unable to validate mvn:org.apache.jclouds.karaf/jclouds-karaf/2.1.0-SNAPSHOT/xml/features
	at org.apache.karaf.features.internal.FeatureValidationUtil.validate(FeatureValidationUtil.java:71)
	at org.apache.karaf.features.internal.FeaturesServiceImpl.validateRepository(FeaturesServiceImpl.java:216)
	at org.apache.karaf.features.internal.FeaturesServiceImpl.internalAddRepository(FeaturesServiceImpl.java:258)
	at org.apache.karaf.features.internal.FeaturesServiceImpl.start(FeaturesServiceImpl.java:1128)
...
Caused by: org.xml.sax.SAXParseException: cvc-elt.1.a: Cannot find the declaration of element 'features'.
	at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source)[:]
	at org.apache.xerces.util.ErrorHandlerWrapper.error(Unknown Source)[:]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)[:]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)[:]
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)[:]
	at org.apache.xerces.impl.xs.XMLSchemaValidator.handleStartElement(Unknown Source)[:]
	at org.apache.xerces.impl.xs.XMLSchemaValidator.startElement(Unknown Source)[:]
	at org.apache.xerces.jaxp.validation.DOMValidatorHelper.beginNode(Unknown Source)[:]
	at org.apache.xerces.jaxp.validation.DOMValidatorHelper.validate(Unknown Source)[:]
	at org.apache.xerces.jaxp.validation.DOMValidatorHelper.validate(Unknown Source)[:]
	at org.apache.xerces.jaxp.validation.ValidatorImpl.validate(Unknown Source)[:]
	at javax.xml.validation.Validator.validate(Unknown Source)[:2.4.0]
	at org.apache.karaf.features.internal.FeatureValidationUtil.validate(FeatureValidationUtil.java:69)
	... 36 more
```

The cause is in the old version of karaf being used. It expects a root element `<features xmlns="http://karaf.apache.org/xmlns/features/v1.0.0"` while our features file has `<features xmlns="http://karaf.apache.org/xmlns/features/v1.4.0"`. It's using a newer namespace which the version of Karaf is not aware of.

Note that jclouds' source feature.xml file is still using the old namespace, but the maven plugin does some processing on it, including changing the version. The generated file gets included in the end.

The changes in this PR don't cause the problem so can be merged independently.

There are two solutions:
  * keep backwards compatibility for the feature files - sticking jclouds-cli to karaf 2.x is a good test case for this
  * update jclouds-cli to current karaf version

jclouds-cli currently depends on Karaf `2.3.11` which is marked as `Not Active` in the [list of supported versions](http://karaf.apache.org/download.html#container). Latest stable version is `4.1.0`.

Thoughts?

**Comment by nacx** [comment]
Nice catch @neykov! I'd say the CLI should always be aligned with this (which was upgraded in the refactor PR):
https://github.com/jclouds/jclouds-karaf/blob/master/pom.xml#L214

**Comment by nacx** [comment]
I actually tried to upgrade the version of karaf in the CLI but all the Console classes used in the CLI runner are deprecated and I have not been able to find proper documentation for their replacement. I tried changing them for the classes I thought could do the same (^^) but no success. I got errors about console classes not being found, and others I don't fully understand.

@neykov @demobox You are more familiar with Karaf than me. Do you think you could have a look at this? Should we ping the dormant karaf guys in our mailing list, or ping the Apache Karaf dev@ list?

**Comment by neykov** [comment]
@nacx you beat me to it. I'll give updating the commands a go.

**Comment by demobox** [comment]
> Should we ping the dormant karaf guys in our mailing list, or ping the Apache Karaf dev@ list?

@nacx I suspect the Karaf dev@ list might be a better bet, specifically for finding suitable replacements/alternative approaches for those classes.

@neykov Thanks for having a go at updating the commands!

**Comment by nacx** [comment]
As part of the recent Karaf fixes we already changed the assembly pom to reference the super-features. Can we close this PR?

**Pull request closed without merging**